### PR TITLE
Use custom abseil with bazel

### LINKED ISSUE #1065: Failed to build with external abseil-cpp
It's simple to understand so I didn't follow the template.

The opentelemetry-cpp vendor parts of the absl source code within the API but without change the namespace. If I build a cpp target included both opentelementry-cpp & abseil-cpp, it would occurred an error (build with bazel & llvm-13):

```text
external/com_google_absl/absl/meta/type_traits.h:162:8: error: redefinition of 'is_detected_convertible_impl<typename std::enable_if<std::is_convertible<Op<Args...>, To>::value>::type, To, Op, Args...>'
struct is_detected_convertible_impl<
       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
bazel-out/k8-fastbuild/bin/external/io_opentelemetry_cpp/api/_virtual_includes/api/opentelemetry/nostd/./absl/types/../types/internal/../../meta/type_traits.h:154:8: note: previous definition is here
struct is_detected_convertible_impl<
       ^
```

The critical rule is to use an alternative namespace than the original namespace `absl` when vendoring any source code.

### PR DESCRIPTION
**Pull request by owent**
Fixes #1065

## Changes

+ Rename `api/include/opentelemetry/nostd/absl` to `api/include/opentelemetry/nostd/internal/absl`
+ Add `--//api:with_abseil=true` to use external abseil when use bazel to build.
+ Add `bazel_with_abseil` into `.github/workflows/ci.yml` to test building with external abseil.

For significant contributions please make sure you have completed the following items:

* [x] `CHANGELOG.md` updated for non-trivial changes
* [x] Unit tests have been added
* [ ] Changes in public API reviewed

### CODE REVIEW

File: bazel/repository.bzl [hunk 1]
```diff
@@ -28,6 +28,17 @@ def opentelemetry_cpp_deps():
         ],
     )
 
+    # Load abseil dependency(optional)
+    maybe(
+        http_archive,
+        name = "com_google_absl",
+        sha256 = "59b862f50e710277f8ede96f083a5bb8d7c9595376146838b9580be90374ee1f",
+        strip_prefix = "abseil-cpp-20210324.2",
```
Comment (esigo) [inline bazel/repository.bzl:36]: abseil version doesn't match with what we use in CI: [20210324.0](https://github.com/owent-contrib/opentelemetry-cpp/blob/3a8157649c8ab0fb147252ce65ecce8ae4046b9c/ci/install_abseil.sh#L11)

Comment (owent) [reply bazel/repository.bzl]: Fixed. The version of internal abseil in `api/include/opentelemetry/nostd/absl` is `v2020-03-03#8` ,should it be upgraded?

Comment (lalitb) [reply bazel/repository.bzl]: > The version of internal abseil in api/include/opentelemetry/nostd/absl is v2020-03-03#8, should it be upgraded?

One of the concerns I see with the periodic upgrade of the internal `nostd::variant` snapshot would be the ABI incompatibility issue if the instrumented library and SDK are built using different otel-cpp releases (with different `nostd::variant` snapshots).

File: api/include/opentelemetry/nostd/variant.h [hunk 2]
```diff
@@ -47,9 +47,9 @@ OPENTELEMETRY_END_NAMESPACE
 #  endif
 
 #  ifdef HAVE_ABSEIL
-#    include <absl/types/variant.h>
+#    include "absl/types/variant.h"
 #  else
-#    include "./absl/types/variant.h"
+#    include "./internal/absl/types/variant.h"
```
[the remark this answers is not in the document]
Comment (lalitb) [reply api/include/opentelemetry/nostd/variant.h:52]: Ok, Thanks for explaining. looks like there is already a bug open for this issue: https://github.com/abseil/abseil-cpp/issues/740

File: api/BUILD [hunk 3]
```diff
@@ -12,11 +12,38 @@
...
+        "//conditions:default": [],
+    }),
     strip_include_prefix = "include",
     tags = ["api"],
+    deps = select({
+        ":with_external_abseil": [
+            "@com_google_absl//absl/base",
+            "@com_google_absl//absl/types:any",
+            "@com_google_absl//absl/types:variant",
+            "@com_google_absl//absl/types:span",
```
Comment (lalitb) [inline api/BUILD:38]: nit - do we need to add a dependency for absl::span?

**Review by lalitb — APPROVED** [review]
LGTM. Thanks for fixing this issue.

File: api/BUILD (hunk 3 above)
Comment (owent) [reply api/BUILD]: `@com_google_absl//absl/types:any` and `@com_google_absl//absl/types:span` are removed. The failed jobs seem not be the problem of this PR.

**Review by ThomsonTan — APPROVED** [review]

**Pull request merged**